### DISCUSSION TITLE
Twitch integration via EventSub

### DISCUSSION THREAD

**Discussion by nuernbergerA** [christophrumpel/larastreamers#128] [Ideas]
Streaming on Twitch is totally different concept than streaming via YouTube.

On YouTube you can schedule a single stream with a starting time.
Twitch instead only allows you to add a weekly schedule due to its nature, it's like a TV show.

The only beneficial way to integrate Twitch for me would be via [EventSub](https://dev.twitch.tv/docs/eventsub).

Workflow:
- Register a twitch streamer
- "Subscribe" channel via API (registers a webhook that gets pinged if the channel is live)
- Decide if we show the stream or not (some tag must be present or something else)

I have already a working API - will create a draft PR

**Comment by christophrumpel** [comment]
Thanks @nuernbergerA,

we discussed this already a bit in #11.
I see currently two issues:

1. we have to subscribe to all streams of a channel (tag can help from the streamer view, but we cannot check the stream if it is relevant from the app view)
2. when we have streams from twitch we need to handle two platforms, like what if a stream is streamed on both platforms, etc.

I don't see them as big issues but they need to be handled and discussed too. I definitely want to support Twitch too somehow.

**Comment by christophrumpel** [comment]
@nuernbergerA, @DanielHe4rt  mentioned these new API endpoints: https://discuss.dev.twitch.tv/t/announcing-api-endpoints-for-twitch-channel-stream-schedules/32137

This will make it now much easier. Challenges I still see:

- How do we know the stream is about Laravel? Is there a way to tag streams on Twitch with Laravel?
- Check if a stream is on both YouTube and Twitch (like mines). Wee need to compare given streams and merge them if they are the same streams, but on different channels

**Comment by austenc** [comment]
Please forgive me if I'm out of the loop on how this app works, but another option may be to allow submission individual streams with a generic URL instead of trying to integrate with twitch directly. 

I'm going to stream on Twitch often (nearly always Laravel), but would be happy to manually submit a few "bigger" stream events as I plan them.

**Comment by christophrumpel** [comment]
Hey, thanks for getting in touch @austenc. What kind of "generic URL" are you thinking of?

Currently, we are using YouTube stream ids for submission. Through these ids we are able to collect all the important stream info like name, date, image, etc.

**Comment by austenc** [comment]
Ah, yeah, I admittedly (sorry!) didn't look at the source to see how that was happening.

Although it works great for Youtube, it may be difficult to incorporate for Twitch given that nothing is really "scheduled". I see you posted about the stream schedule API that Twitch has now... problem is, what if you miss a stream? 

Anyway, I was thinking of a form that would allow you to fill out all those fields manually. I thought that's how it was working already, but I assumed wrong! Totally understand if supporting something like that isn't something you want to do as a maintainer, though. Thanks!

**Comment by aarondfrancis** [comment]
I'd be totally fine to just manually add my streams vs having a full on integration with Twitch! Is that doable?

**Reply by christophrumpel** [reply]
The whole app is currently focused on YouTube, so there is no `easy way` to add Twitch. But since more and more people ask about it, and the Twitch API around a schedule looks much better than in the past, I will give it another look to find the easiest way to support it. I will give it a closer look after Laracon Online and share my findings. Would also be fun to stream about adding Twitch support since this would be a big PR. Thanks for the feedback.

**Reply by skylerkatz** [reply]
As I mentioned on Twitter, I have some experience with the Twitch API, and live streaming in general, and would be happy to assist with this feature.

While the auto-import channel features of the app may not be suitable for an initial implementation, we could have them manually submit their channel and either the title of their stream, or the "segment id" which is unique per event. If they enter the title we would need to look up the schedule and match the title with the _next_ upcoming event, or we would also need to ask for a start date to match things up. Grabbing the Segment Id would be the easiest thing for us, but it is a little difficult for the user to find that without doing a bit of hackery.

Future improvements would be to auto import channels that have a schedule and are tagged with something like `laravel` or `larastreamers`.

**Reply by christophrumpel** [reply]
Thanks a lot Skyler, yeah the auto-import solution is not necessary right now. Perfect would be if the user could provide some kind of stream id like YouTube, but from what I remember, this is different with Twitch. What is the segment id on how does a streamer get it? Entering a title and looking for something similar from the streamer sounds a bit error-prone.

**Reply by skylerkatz** [reply]
If you consider a stream on twitch to be never ending, the segment id is a unique identifier for a particular start and end time. For a scheduled stream on twitch it would have a start and end time, but if you just "go live" it has a start but no end time.

Given a channel name, and a start time, we can find the segment id of a pre-scheduled stream.

**Reply by christophrumpel** [reply]
ah ok, interesting approach. Ok, let's say the streamer gives us enough information to make a call to his schedule and find the right "segement"/stream. (from here? https://dev.twitch.tv/docs/api/reference/#get-channel-stream-schedule)

I guess we then have enough info about the segment to put in on Larastreamers. How would we check then if this segment is live?
Over the streamer's channel here? https://dev.twitch.tv/docs/api/reference#get-streams

**Reply by skylerkatz** [reply]
If you are trying to make sure that the segment is _actually live_ then I think we would want to subscribe to the `stream.online` event for the user. We can do things like subscribe to the event from the time the stream is submitted to larastreamers until the date passes, or just continue to be subscribed but ignore webhooks when the stream was not scheduled in larastreamers.

**Reply by christophrumpel** [reply]
Started looking into this a bit more, @skylerkatz do I understand this correctly that I can only subscribe to if a `streamer` is online and `not` if a stream/segment is online?

**Reply by skylerkatz** [reply]
That is correct.

**Comment by christophrumpel** [comment]
Alright, I tried to sum up the flows of how working with Twitch streams could work. What do you think?

**The scenario where we store a whole twitch channel:**

1. We add `channel name` to backend-form
2. We use the Twitch  `users` endpoint with the `login channel name` to get the `broadcaster id` of the channel
3. We use the Twitch `schedule` endpoint to get the schedule of a channel with the broadcaster id
4. We `store` all the schedule segments
5. We use the Twitch `eventsub/subscriptions` endpoint to set up a subscription to the `stream.online` and `stream.offline` event for the current broadcaster
6. When we receive `stream.online` event, we check if we have this stream, if not we add it and set it to live, if yes, we set the given stream to `live`
7. When we receive the `stream.offline` event, we check if we have this stream, if yes, we set the stream to `finished` 
8. `Regularly check the schedule` of the broadcaster for new streams or updates

**The scenario where the user can submit a twitch stream:**

1. User provides channel name and stream title
2. We use the Twitch  `users` endpoint with the login channel name to get the broadcaster id of the channel
3. We use the Twitch `schedule` endpoint to get the schedule of a channel with the broadcaster id
4. We look through the `segments` to find the right one with the correct name
5. We store the `segment` as stream
6. We use the Twitch `eventsub/subscriptions` endpoint to set up a subscription to the `stream.online` and `stream.offline` event for the current broadcaster
7. When we receive `stream.online`, we check if we have this stream, if not we skip it, if yes, set this stream to `live`
8. When we receive the `stream.offline` event, we check if we have this stream, if not we skip it, if yes, we set the stream to `finished` and cancel the subscription through the `eventsub/subscriptions` endpoint
9. `Regularly check the schedule` of the broadcaster to see if given streams have changed, if yes, update them

**Reply by skylerkatz** [reply]
Not sure if it makes sense for an initial implementation, but if you submit your whole channel it might be nice to say "subscribe to all segments on my channel with a tag of X" where X is a form field they fill out. If blank, subscribe to everything. If filled, skip segments that don't include X.

**Reply by christophrumpel** [reply]
And "x" would be something in the name? Or some kind of tag?

**Reply by skylerkatz** [reply]
yeah, X would be a tag that the user enters when subscribing their channel.

For example,

`#laravel` or `#public` whatever the user chooses.

**Comment by christophrumpel** [comment]
@skylerkatz I just noticed (during stream) that for the `online` / `offline` events, we don't get any info about the "stream/segment", this is only about the channel. So I have no way to connect `online` to a specific stored stream, I can just guess it is the nearest one... at least it seems so after checking the API again. Any idea?

**Reply by christophrumpel** [reply]
I did some more research, and then talked on the Twitch forum about it: https://discuss.dev.twitch.tv/t/how-to-know-which-stream-segment-is-live/41800/7

It seems it is almost impossible to tie an twitch scheduled stream, to when a channel is live. In the Twitch world there is no real connection between a channel which is live and a scheduled stream. And there is NO "stream that is live". It's always a channel that is live.

**Reply by skylerkatz** [reply]
Sorry I did not respond yesterday. The only recommendation would be to match the current channel title with the segment title you expect.

Another option is to tell the person syncing their schedule that they need to include a #larastream tag on segments that they want to be on the site
